pallets/sto/src/lib.rs

```diff
@@ -0,0 +1,398 @@
+
+        /// Purchase tokens from an ongoing offering.
+        #[weight = 2_000_000_000]
+        pub fn invest(
+            origin,
+            investment_portfolio: PortfolioId,
+            funding_portfolio: PortfolioId,
+            offering_asset: Ticker,
+            fundraiser_id: u64,
+            investment_amount: T::Balance,
+            max_price: T::Balance,
+            reciept: Option<ReceiptDetails<T::AccountId, T::OffChainSignature>>
+        ) -> DispatchResult {
+            let sender = ensure_signed(origin.clone())?;
+            let did = Context::current_identity_or::<Identity<T>>(&sender)?;
+
+            ensure!(investment_portfolio.did == did, Error::<T>::InvalidPortfolio);
+            ensure!(funding_portfolio.did == did, Error::<T>::InvalidPortfolio);
+
+            let now = Timestamp::<T>::get();
+            let fundraiser = <Fundraisers<T>>::get(offering_asset, fundraiser_id).ok_or(Error::<T>::FundraiserNotFound)?;
+            ensure!(fundraiser.start <= now, Error::<T>::FundraiserNotStarted);
+            if let Some(end) = fundraiser.end {
+                ensure!(end > now, Error::<T>::FundraiserExpired);
+            }
+
+            // Remaining tokens to fulfil the investment amount
+            let mut remaining = investment_amount;
+            // Total cost to to fulfil the investment amount.
+            // Primary use is to calculate the blended price (offering_token_amount / cost).
+            // Blended price must be <= to max_price or the investment will fail.
+            let mut cost = T::Balance::from(0);
+            // Individual purchases from each tier that accumulate to fulfil the investment amount.
+            // Tuple of (tier_id, amount to purchase from that tier).
+            let mut purchases = Vec::new();
+
+            for (id, tier) in fundraiser.tiers.iter().enumerate() {
+                // fulfilled the investment amount
+                if remaining == 0.into() {
+                    break
+                }
+
+                // tier is exhausted, move on
+                if tier.remaining == 0.into() {
+                    continue
+                }
+
+                // Check if this tier can fulfil the remaining investment amount.
+                // If it can, purchase the remaining amount.
+                // If it can't, purchase what's remaining in the tier.
+                let purchase_amount = if tier.remaining >= remaining {
+                    remaining
+                } else {
+                    tier.remaining
+                };
+
+                remaining -= purchase_amount;
+                purchases.push((id, purchase_amount));
+                cost = cost.checked_add(
+                    &remaining
+                    .checked_mul(&tier.price)
+                    .ok_or(Error::<T>::Overflow)?
+                ).ok_or(Error::<T>::Overflow)?;
+            }
+
+            ensure!(remaining == 0.into(), Error::<T>::InsufficientTokensRemaining);
+
```

Comment:
Need a check for `max_price`. Maybe replace `max_price` with `max_cost` to make it a direct comparison.